### PR TITLE
Changed CI to Nix

### PR DESCRIPTION
**Pull request by mrtuxa**
Nix

### CODE REVIEW

File: .github/workflows/rust.yaml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      - name: Lint
+        run: nix-shell -p cargo openssl pkg-config -I nixpkgs=channel:nixos-unstable --command "cargo clippy --all-targets --all-features -- -D warnings"
+
+  fmt:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: cachix/install-nix-action@v20
+      - name: Format check
+        run: nix-shell -p cargo openssl pkg-config -I nixpkgs=channel:nixos-unstable --command "cargo fmt -- --check"
```
Comment (aboehm) [inline .github/workflows/rust.yaml:61]: Why unstable?

File: .github/workflows/rust.yaml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            result/bin/spaceapi-dezentrale-client
+            result/bin/spaceapi-dezentrale-server
+
+  lint:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: cachix/install-nix-action@v20
+      - name: Lint
+        run: nix-shell -p cargo openssl pkg-config -I nixpkgs=channel:nixos-unstable --command "cargo clippy --all-targets --all-features -- -D warnings"
```
Comment (aboehm) [inline .github/workflows/rust.yaml:53]: Why unstable?

File: .github/workflows/rust.yaml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      - name: Build and test
+        run: nix-shell -p cargo rustc openssl pkg-config -I nixpkgs=channel:nixos-unstable --command "cargo test"
+
+  binaries_x86_64:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: cachix/install-nix-action@v20
+        with:
+          github_access_token: ${{ secrets.GITHUB_TOKEN }}
```
Comment (aboehm) [inline .github/workflows/rust.yaml:27]: Why leaking the secret token?

File: .github/workflows/rust.yaml [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: cachix/install-nix-action@v20
+        with:
+          nix_path: nixpkgs=channel:nixos-unstable
+      - name: Build and test
+        run: nix-shell -p cargo rustc openssl pkg-config -I nixpkgs=channel:nixos-unstable --command "cargo test"
```
Comment (aboehm) [inline .github/workflows/rust.yaml:19]: Why unstable?

File: Cargo.lock [hunk 5]
```diff
@@ -0,0 +1,2030 @@
+# This file is automatically @generated by Cargo.
```
Comment (aboehm) [inline Cargo.lock:1]: Why did you add the lock file?

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -1,48 +0,0 @@
...
-on:
-  push:
-    branches: [ "main" ]
-  pull_request:
-    branches: [ "main" ]
-
-jobs:
-  test:
-    runs-on: ubuntu-22.04
-    container: rust:1.67
```
Comment (aboehm) [inline .github/workflows/ci.yml:11]: Why did you removed the explicit version setting?

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -1,48 +0,0 @@
-name: CI
-on:
-  push:
-    branches: [ "main" ]
-  pull_request:
-    branches: [ "main" ]
-
-jobs:
-  test:
-    runs-on: ubuntu-22.04
```
Comment (aboehm) [inline .github/workflows/ci.yml:10]: Why did you removed the explicit target version setting?

**Review by aboehm — CHANGES_REQUESTED** [review]
This changes will add more complexity and build time, why?

File: Cargo.lock (hunk 5 above)
Comment (mrtuxa) [reply Cargo.lock:1]: Because Nix needs it to build it

File: .github/workflows/rust.yaml (hunk 3 above)
Comment (mrtuxa) [reply .github/workflows/rust.yaml:27]: It does not leak the token

File: .github/workflows/rust.yaml (hunk 2 above)
Comment (mrtuxa) [reply .github/workflows/rust.yaml:53]: I can change it to stable if needed

**Comment by mrtuxa** [comment]
> This changes will add more complexity and build time, why?

More complexity? It will help to easy share the build env

File: .github/workflows/ci.yml (hunk 7 above)
Comment (mrtuxa) [reply .github/workflows/ci.yml:10]: I can define if needed

File: .github/workflows/rust.yaml (hunk 3 above)
Comment (aboehm) [reply .github/workflows/rust.yaml:27]: Why is the usage of a secret token required, when it work without it with the existing toolchain?

File: Cargo.lock (hunk 5 above)
Comment (aboehm) [reply Cargo.lock:1]: I think nix is a package manager. Why does introduction of nix change the behaviour of cargo?

File: .github/workflows/rust.yaml (hunk 3 above)
Comment (mrtuxa) [reply .github/workflows/rust.yaml:27]: `github_access_token`: configure Nix to pull from GitHub using the given GitHub token. This helps work around rate limit issues. Has no effect when `access-tokens` is also specified in `extra_nix_config`.

**Comment by aboehm** [comment]
It's already portable. It uses the recommended build method via container and it's rustup compatible. If you like to add a nix support, please add a optional workflow.

File: Cargo.lock (hunk 5 above)
Comment (mrtuxa) [reply Cargo.lock:1]: It's the behavior or cargo2nix it needs the cargo lock file to build it

**Comment by mrtuxa** [comment]
> It's already portable. It uses the recommended build method via container and it's rustup compatible. If you like to add a nix support, please add a optional workflow.

It unnecessary to add a second workflow. However, it is also worth noting that adding support for Nix can simplify the build environment and save resources such as the Rust compiler, among others. Nix is known for its functional approach to package management and can provide additional benefits to users who prefer this method.

File: Cargo.lock (hunk 5 above)
Comment (aboehm) [reply Cargo.lock:1]: Why another tool?

Comment (mrtuxa) [reply Cargo.lock:1]: > Why another tool?

Cargo2nix is a tool that allows you to generate Nix packages from Rust projects built with Cargo. By using cargo2nix as an additional tool, you can easily create Nix expressions for your Rust projects and manage their dependencies in a declarative and reproducible way.

**Comment by mrtuxa** [comment]
> > Nix is known for its functional approach to package management and can provide additional benefits to users who prefer this method.
> 
> You give the argument for the optional workflow: Your current implementation add more resources (4.5 min before vs. >8 min to build step). Please add a developer documentation for building it the nix way.




However, Nix has many advantages, such as reproducibility of builds, isolation of dependencies, and support for multiple versions of libraries, which can lead to a stable and consistent build process.

**Comment by mrtuxa** [comment]
> Please argument on the points.
> 
> We don't have multiple versions of libraries at the moment. We don't have a unstable build process. All instabilities in the build process in the past are based on the use of outdated rust versions, which can be solved via a [toolchain file](https://rust-lang.github.io/rustup/overrides.html#the-toolchain-file). This will use the common rust tools and adds no additional tools to the stack.
> 
> We have a running build pipeline, which have a small dependency footprint and is faster than the submitted changes. The changes adds more technical debts (tools, requirements, commands wrappers), which should have a good explanation why to do this.
> 
> We can add this feature as an optional one as recommended before. We shouldn't replace a common approach with a specialized one. People who don't us nix should be able to develop and understand the pipeline without the knowledge of nix.


 
I understand your points, but I still believe that using Nix as the preferred build manager can bring significant benefits to the project.

Regarding the multiple versions of libraries, while you may not currently have this requirement, it's still a potential issue that can arise in the future. Nix allows for easy management of different versions of dependencies and can prevent version conflicts that can lead to unstable builds.

Regarding the build process, while you may have a running build pipeline currently, there's always room for improvement. Using Nix can potentially improve build times and ensure consistent builds across different machines and environments.

In terms of technical debt, while adding new tools and requirements can increase complexity initially, it can also improve the maintainability and stability of the project in the long run. And providing clear documentation on the use of Nix can mitigate the learning curve for developers.

Lastly, I agree that the Nix workflow can be optional, but I still believe that promoting its use as the preferred approach can benefit the project as a whole. Developers who don't use Nix can still use the pipeline without any knowledge of Nix, but those who do can take advantage of its benefits.

**Comment by aboehm** [comment]
Ok, then we have a solution. Make the pipeline optional and add all required changes for that.
